### ISSUE TITLE
Enhance when, choose, map and repeat to accept promises

### ISSUE THREAD

**Issue by claytongulick** [lit/lit#3447]
### Should this be an RFC?

- [X] This is not a substantial change

### Which package is this a feature request for?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

It would be a great addition for when, choose, map and repeat to accept promises as well as literals for their parameters, and only render the template once the promise has resolved.

This would be an amazing feature - especially for those of us who use lit-html directly and manage the render cycle manually.

### Alternatives and Workarounds

I don't know of a great way to achieve this other than reproducing the functionality of those directives and using "until" or waiting for a promise to resolve before rendering.

**Comment by justinfagnani** [comment]
You can already do this with until and one of the helpers:

```ts
html`
  <div>${until(somePromise.then((value) =>
    when(value === whatever, () => 
      html`<div>${value}</div>`)
  }</div>`
```

The core problem of doing this in the conditional helpers is that the conditional directives don't actually take a value, they take a _condition_. If you want to evaluate that condition on a promised value, you have to have wait for the promises before condition evaluation. That's what `until()` already does.

Nesting directive like that can get kind of messy, and the conditional directive and just small wrappers around standard JS control flow - they only exist to make some simple cases format nicer if you like that - so I'd probably recommend dropping more into JS if you have complicated async control flow so that you can use async functions:

```ts
html`<div>${until((async () => {
  const value = await somePromise;
  if (value === whatever) {
    return html`<div>${value}</div>`;
  }
})())}</div>`;
```

Now we could potentially remove some of the symbol noise by letting `until()` take no-arg async functions directly.

But then this brings two another problem with async directives:
1. They make template hard to reason about. You don't know when a template is "done" or if it ever might render again.
2. They don't integrate with the host element lifecycle. Other state could be out of sync by the time the until'ed template partial renders, and things like `updateComplete` don't include the async template state.
3. They sometimes more difficult to integrate with change detection of bindings. Often API calls will produce new Promises wrapping the previous value seen at a binding. In these cases you do work to wait on the promise every render. It's better in these cases to pull the promise out of the render loop.

This is why we're moving more towards recommending reactive controllers to handle async state. The template should be sync and based on the current snapshot of state, then outside of the template async tasks should update that state snapshop and trigger a re-render. Things like our Task controller basically turn async state into sync state for rendering, but then trigger a re-render when state changes. This pattern makes it a lot easier to reason about templates.

If you're managing lifecycle manually I would still recommend a pattern where async tasks just trigger a new render. Re-renders are fast.

**Comment by claytongulick** [comment]
@justinfagnani yep, thanks! I'm doing exactly what you're suggesting, my request comes from "de-noising" templates, as you say.

Would be super convenient to be able to stick promises in those directives and have them smart enough to await them all, would save on a good bit of boilerplate code.

Currently, I inherit directly from HTMLElement and call render() from property setters, those get invoked as the result of an async call. Pretty straightforward reactive code.

Being able to do something like:
```
${ when(user.isAuthenticated(), this.showSomething, this.showLogin)}
```

and especially
```
${map(this.fetchData(), (row) => html`...`)}
```

would just be really nice :D
